Limit the amount of data that PNGImageDecoder may decode

aperezdc opened the pull request:
#### 7837212f52f33db3d22f69704ed7654aedd313e8
<pre>
Limit the amount of data that PNGImageDecoder may decode
<a href="https://bugs.webkit.org/show_bug.cgi?id=302941">https://bugs.webkit.org/show_bug.cgi?id=302941</a>

Reviewed by Patrick Griffis.

Keep a running count of the amount of pixels decoded, update it on every
frame decoded for an image, and bail out if it would grow large enough
to potentially cause memory exhaustion. While at it, limit the maximum
amount of frames that an APNG file may have, and define it to be large
enough to cover ~70 minutes of animation at 60fps.

* Source/WebCore/platform/image-decoders/png/PNGImageDecoder.cpp:
(WebCore::PNGImageDecoder::headerAvailable):
(WebCore::PNGImageDecoder::readChunks):
* Source/WebCore/platform/image-decoders/png/PNGImageDecoder.h:

Canonical link: <a href="https://commits.webkit.org/303448@main">https://commits.webkit.org/303448@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cbfe5df6e2c7cf386e575bbc49c80e8bc05f77e0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132442 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4937 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43493 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139957 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84404 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134312 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4981 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4696 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101252 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135388 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3505 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118639 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82045 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1228 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83187 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112240 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36756 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142611 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4607 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37345 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109628 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4689 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3976 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109809 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27825 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3491 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114913 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57921 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4661 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33263 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4495 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68112 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4752 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4618 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/image-decoders/png/PNGImageDecoder.cpp:
```diff
@@ -66,8 +66,11 @@ const double cMaxGamma = 21474.83;
-// Protect against large PNGs. See Mozilla's bug #251381 for more info.
-const unsigned long cMaxPNGSize = 1000000UL;
+// Protect against APNGs with huge amounts of frames, and PNGs with large amounts
+// of pixel data un general. See bug #302220 and Mozilla's bug #251381 for more info.
```
```suggestion
// of pixel data in general. See bug #302220 and Mozilla's bug #251381 for more info.
```

mcatanzaro commented on Source/WebCore/platform/image-decoders/png/PNGImageDecoder.cpp:
```diff
@@ -66,8 +66,11 @@ const double cMaxGamma = 21474.83;
+// Protect against APNGs with huge amounts of frames, and PNGs with large amounts
+// of pixel data un general. See bug #302220 and Mozilla's bug #251381 for more info.
+static constexpr unsigned long cMaxFrameCount = (1 << 18) - 1;
+static constexpr unsigned long cMaxPNGSize = 1000000UL;
```
Let's not mess around with silly types like "long":

```suggestion
static constexpr uint32_t cMaxFrameCount = (1 << 18) - 1;
static constexpr uint32_t cMaxPNGSize = 1000000;
```

mcatanzaro commented on Source/WebCore/platform/image-decoders/png/PNGImageDecoder.cpp:
```diff
@@ -283,10 +286,12 @@ void PNGImageDecoder::headerAvailable()
 
     // Protect against large images.
-    if (width > cMaxPNGSize || height > cMaxPNGSize) {
+    const auto pixelCount = checkedSum<size_t>(width, height, m_decodedPixelCount);
```
I don't think we need to write `const` everywhere just because we can. Use it where it adds value. In a local variable, is it really adding value? Same comment applies for the `const auto pixelCount` below, as well.

aperezdc replied:
The intention is that we would never change the value of `pixelCount` after calculating it, so I would rather have the compiler complain if we accidentally did that, even if the generated code is the same.

mcatanzaro commented on Source/WebCore/platform/image-decoders/png/PNGImageDecoder.h:
```diff
@@ -112,6 +112,7 @@ namespace WebCore {
         bool m_hasInfo;
         int m_gamma;
         size_t m_frameCount;
+        size_t m_decodedPixelCount { 0 };
```
Style nit: yes this is the better way to initialize member variables, but it's also weird for it to be the _only_ variable initialized here rather than in the constructors. I would have preferred to match the existing code.

aperezdc replied:
Sure, I'll move the initialization to the constructor for the sake of coherency.

TingPing reviewed (approved):
> define it to be large enough to cover ~70 minutes of animation at 60fps.

Maybe this comment should be in the code next to the number for context.

aperezdc committed:
Limit the amount of data that PNGImageDecoder may decode
https://bugs.webkit.org/show_bug.cgi?id=302941

Reviewed by Patrick Griffis.

Keep a running count of the amount of pixels decoded, update it on every
frame decoded for an image, and bail out if it would grow large enough
to potentially cause memory exhaustion. While at it, limit the maximum
amount of frames that an APNG file may have, and define it to be large
enough to cover ~70 minutes of animation at 60fps.

* Source/WebCore/platform/image-decoders/png/PNGImageDecoder.cpp:
(WebCore::PNGImageDecoder::headerAvailable):
(WebCore::PNGImageDecoder::readChunks):
* Source/WebCore/platform/image-decoders/png/PNGImageDecoder.h:

Canonical link: https://commits.webkit.org/303448@main
